pkg/remote/codec/unknown/unknown.go

```diff
@@ -0,0 +1,286 @@
+func NewUnknownCodec(code remote.PayloadCodec) remote.PayloadCodec {
+	return &unknownCodec{code}
+}
+
+// Marshal implements the remote.PayloadCodec interface.
+func (c unknownCodec) Marshal(ctx context.Context, msg remote.Message, out remote.ByteBuffer) error {
+	ink := msg.RPCInfo().Invocation()
+	data := msg.Data()
+
+	res, ok := data.(*unknowns.Result)
+	if !ok {
+		return c.Codec.Marshal(ctx, msg, out)
+	}
+	if len(res.Success) == 0 {
+		return errors.New("unknown messages cannot be empty")
+	}
+	if msg.MessageType() == remote.Exception {
+		return c.Codec.Marshal(ctx, msg, out)
+	}
+	if ink, ok := ink.(rpcinfo.InvocationSetter); ok {
+		ink.SetMethodName(res.Method)
+		ink.SetServiceName(res.ServiceName)
+	} else {
+		return errors.New("the interface Invocation doesn't implement InvocationSetter")
+	}
+	if err := encode(res, msg, out); err != nil {
+		return c.Codec.Marshal(ctx, msg, out)
+	}
+	return nil
+}
+
+// Unmarshal implements the remote.PayloadCodec interface.
+func (c unknownCodec) Unmarshal(ctx context.Context, message remote.Message, in remote.ByteBuffer) error {
+	ink := message.RPCInfo().Invocation()
+	service, method, size, err := decode(message, in)
+	if err != nil {
+		return c.Codec.Unmarshal(ctx, message, in)
+	}
+	err = codec.SetOrCheckMethodName(method, message)
+	if te, ok := err.(*remote.TransError); ok && te.TypeID() == remote.UnknownMethod {
+		svcInfo, err := message.SpecifyServiceInfo(unknowns.UnknownService, unknowns.UnknownMethod)
+		if err != nil {
+			return err
+		}
+
+		if ink, ok := ink.(rpcinfo.InvocationSetter); ok {
+			ink.SetMethodName(unknowns.UnknownMethod)
+			ink.SetPackageName(svcInfo.GetPackageName())
+			ink.SetServiceName(unknowns.UnknownService)
+		} else {
+			return errors.New("the interface Invocation doesn't implement InvocationSetter")
+		}
+		if err = codec.NewDataIfNeeded(unknowns.UnknownMethod, message); err != nil {
+			return err
+		}
+
+		data := message.Data()
+
+		if data, ok := data.(*unknowns.Args); ok {
+			data.Method = method
+			data.ServiceName = service
+			err := in.Skip(int(size))
+			if err != nil {
+				return err
+			}
+			buf, err := in.Next(in.ReadableLen())
+			if err != nil {
+				return err
+			}
+			data.Request = buf
+		}
+		return nil
+	}
+
+	return c.Codec.Unmarshal(ctx, message, in)
+}
+
+// Name implements the remote.PayloadCodec interface.
+func (c unknownCodec) Name() string {
+	return "unknownMethodCodec"
+}
+
+func write(dst, src []byte) {
+	copy(dst, src)
+}
+
+func decode(message remote.Message, in remote.ByteBuffer) (string, string, int32, error) {
+	code := message.ProtocolInfo().CodecType
+	if code == serviceinfo.Thrift {
+		return decodeThrift(message, in)
+	} else if code == serviceinfo.Protobuf {
+		return decodeProtobuf(message, in)
+	}
+	return "", "", 0, nil
+}
+
+// decodeThrift  Thrift decoder
+func decodeThrift(message remote.Message, in remote.ByteBuffer) (string, string, int32, error) {
```

Comment:
decodeThrift and decodeProtobuf is similar.
Why split it into two?

Reply:
Peek read data is partially different, so it is separated